### DISCUSSION TITLE
Doc for "keep_cache" doesn't seem true

### DISCUSSION THREAD

**Discussion by raj-prince** [libfuse/libfuse#947] [General]
```
/** Can be filled in by open. It signals the kernel that any
	    currently cached file data (ie., data that the filesystem
	    provided the last time the file was open) need not be
	    invalidated. Has no effect when set in other contexts (in
	    particular it does nothing when set by opendir()). */
```
Src: https://github.com/libfuse/libfuse/blob/1552b467fcd7751360299c5139382d78538e12b3/include/fuse_common.h#L58

Comment explicitly says, this flag doesn't do anything when set by opendir(), but it doesn't seem true since it works for me.

Also, found the handling in the kernel fuse driver: https://github.com/torvalds/linux/blob/dccb07f2914cdab2ac3a5b6c98406f765acab803/fs/fuse/dir.c#L712

Could you please someone validate, as I am new to these code-bases? So, not sure if I am missing something.

**Comment by raj-prince** [comment]
Another `typo` in the description of this commit: https://github.com/torvalds/linux/commit/6433b8998a21dc597002731c4ceb4144e856edc4

So FOPEN_KEEP_CACHE and **FOPEN_KEEP_CACHE** flags should be used together to
make use of the directory caching facility introduced in the following
patches.

Seem, author is trying to write:
`So FOPEN_KEEP_CACHE and FOPEN_CACHE_DIR flags should be used together to
make use of the directory caching facility introduced in the following
patches.`

**Comment by bsbernd** [comment]
I'm not sure what you mean, the kernel function you are looking at is `fuse_create_open`and that is never called for directories. The corresponding flag for directories is FOPEN_CACHE_DIR (struct fuse_file_info::cache_readdir). 
So no, the comment is correct.

**Reply by raj-prince** [reply]
Yes, I agree. By looking at the code, it seems `FOPEN_KEEP_CACHE` doesn't do anything for `opendir()`.

**Comment by bsbernd** [comment]
Well maybe the commit message says  that, but code is actually different - it either *checks* for FOPEN_CACHE_DIR or FOPEN_KEEP_CACHE.

**Reply by raj-prince** [reply]
If the above is not true, what's the other way of invalidating readdir() cache?

My understanding: after playing with these flags in different fuse (jacobsa/fuse) library.
```
#define FOPEN_DIRECT_IO		(1 << 0)
#define FOPEN_KEEP_CACHE	(1 << 1)
#define FOPEN_CACHE_DIR		(1 << 3)
```

- FOPEN_DIRECT_IO is to control if **file** content should be cached or not.
- FOPEN_CACHE_DIR is to control if **dir** content should be cached or not.
- FOPEN_KEEP_CACHE is to invalidate either of the file or dir content.

I wanted to validate the behavior, that's why I looked into libfuse code.

**Reply by bsbernd** [reply]
I'm not sure what you mean with 
 > If the above is not true, what's the other way of invalidating readdir() cache?
 
And  
- FOPEN_KEEP_CACHE is to  control if file content cache should _not_ be invalidated on file open. I.e. setting it might violate close-to-open coherency.

**Reply by raj-prince** [reply]
`FOPEN_KEEP_CACHE is to control if file content cache should not be invalidated on file open. I.e. setting it might violate close-to-open coherency.` - Yes, that's true.

My question is - how to maintain `close-to-open` consistency for dir content cache?

**Comment by bsbernd** [comment]
close-to-open is default for directories. Question, do you possible have an opendir() method that returns ENOSYS?
